File: examples/integrations/go/achordconnect/achordconnect.go
package achordconnect

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"html"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

type Client struct {
	BaseURL      string
	ClientID     string
	ClientSecret string
	HTTP         *http.Client
}

type User struct {
	ID         string         `json:"id"`
	Name       string         `json:"name"`
	Email      *string        `json:"email,omitempty"`
	Username   *string        `json:"username,omitempty"`
	AvatarURL  *string        `json:"avatarUrl,omitempty"`
	Attributes map[string]any `json:"attributes,omitempty"`
}

type LaunchContext struct {
	Theme        string `json:"theme,omitempty"`
	Locale       string `json:"locale,omitempty"`
	ReturnOrigin string `json:"returnOrigin,omitempty"`
}

type LaunchResponse struct {
	LaunchURL string `json:"launchUrl"`
	ExpiresAt string `json:"expiresAt"`
}

func (c Client) CreateLaunchTicket(
	ctx context.Context,
	user User,
	launchContext LaunchContext,
) (LaunchResponse, error) {
	if strings.TrimSpace(user.ID) == "" {
		return LaunchResponse{}, errors.New("achord connect user ID must be a non-empty string")
	}
	endpoint, err := url.JoinPath(c.BaseURL, "/api/v1/integrations/universal/launch-tickets")
	if err != nil {
		return LaunchResponse{}, err
	}
	body, err := json.Marshal(map[string]any{
		"user":    user,
		"context": launchContext,
	})
	if err != nil {
		return LaunchResponse{}, err
	}
	req, err := http.NewRequestWithContext(
		ctx,
		http.MethodPost,
		endpoint,
		bytes.NewReader(body),
	)
	if err != nil {
		return LaunchResponse{}, err
	}
	req.SetBasicAuth(c.ClientID, c.ClientSecret)
	req.Header.Set("Content-Type", "application/json")
	httpClient := c.HTTP
	if httpClient == nil {
		httpClient = &http.Client{Timeout: 8 * time.Second}
	}
	res, err := httpClient.Do(req)
	if err != nil {
		return LaunchResponse{}, err
	}
	defer res.Body.Close()
	limited, err := io.ReadAll(io.LimitReader(res.Body, 1<<20))
	if err != nil {
		return LaunchResponse{}, err
	}
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return LaunchResponse{}, fmt.Errorf("achord connect HTTP %d", res.StatusCode)
	}
	var envelope struct {
		Data LaunchResponse `json:"data"`
	}
	if err := json.Unmarshal(limited, &envelope); err != nil {
		return LaunchResponse{}, err
	}
	return envelope.Data, nil
}

func IframeHTML(launchURL string, title string) (string, error) {
	parsed, err := url.Parse(launchURL)
	if err != nil || parsed.Host == "" || (parsed.Scheme != "https" && parsed.Scheme != "http") {
		return "", errors.New("achord connect launch URL must be an absolute HTTP or HTTPS URL")
	}
	if title == "" {
		title = "服务请求"
	}
	return fmt.Sprintf(
		`<iframe src="%s" title="%s" style="width:100%%;min-height:720px;border:0" allow="clipboard-write"></iframe>`,
		html.EscapeString(parsed.String()),
		html.EscapeString(title),
	), nil
}

func VerifyWebhook(
	secret string,
	rawBody []byte,
	eventID string,
	timestamp string,
	signature string,
	now time.Time,
) error {
	if eventID == "" || !strings.HasPrefix(signature, "v1=") {
		return errors.New("missing webhook headers")
	}
	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil || abs(now.Unix()-ts) > 300 {
		return errors.New("expired webhook")
	}
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(timestamp + "."))
	_, _ = mac.Write(rawBody)
	expected := mac.Sum(nil)
	actual, err := hex.DecodeString(strings.TrimPrefix(signature, "v1="))
	if err != nil || !hmac.Equal(expected, actual) {
		return errors.New("invalid webhook signature")
	}
	return nil
}

func abs(value int64) int64 {
	if value < 0 {
		return -value
	}
	return value
}

File: examples/integrations/node/achord-connect.ts
import { createHmac, timingSafeEqual } from "node:crypto";

export type AchordConnectUser = {
  id: string;
  name: string;
  email?: string | null;
  username?: string | null;
  avatarUrl?: string | null;
  attributes?: Record<string, string | number | boolean>;
};

export async function createLaunchTicket(input: {
  baseUrl: string;
  clientId: string;
  clientSecret: string;
  user: AchordConnectUser;
  context?: {
    theme?: "light" | "dark" | "system";
    locale?: string;
    returnOrigin?: string;
  };
}) {
  if (typeof input.user.id !== "string" || input.user.id.trim() === "") {
    throw new TypeError("Achord Connect user.id must be a non-empty string");
  }
  const response = await fetch(
    new URL("/api/v1/integrations/universal/launch-tickets", input.baseUrl),
    {
      method: "POST",
      headers: {
        Authorization: `Basic ${Buffer.from(`${input.clientId}:${input.clientSecret}`).toString("base64")}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ user: input.user, context: input.context ?? {} }),
    },
  );
  const payload = (await response.json()) as {
    data?: { launchUrl: string; expiresAt: string };
    error?: { message?: string };
  };
  if (!response.ok || !payload.data) {
    throw new Error(payload.error?.message ?? `Achord Connect HTTP ${response.status}`);
  }
  return payload.data;
}

export function iframeHtml(launchUrl: string, title = "服务请求") {
  const parsed = new URL(launchUrl);
  if (!['https:', 'http:'].includes(parsed.protocol)) {
    throw new TypeError("Achord Connect launchUrl must use HTTP or HTTPS");
  }
  const escapeAttribute = (value: string) =>
    value
      .replace(/&/g, "&amp;")
      .replace(/"/g, "&quot;")
      .replace(/'/g, "&#39;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;");
  const escapedUrl = escapeAttribute(parsed.toString());
  const escapedTitle = escapeAttribute(title);
  return `<iframe src="${escapedUrl}" title="${escapedTitle}" style="width:100%;min-height:720px;border:0" allow="clipboard-write"></iframe>`;
}

export function verifyWebhook(input: {
  secret: string;
  rawBody: string;
  eventId: string | null;
  timestamp: string | null;
  signature: string | null;
  toleranceSeconds?: number;
  nowSeconds?: number;
}) {
  if (!input.eventId || !input.timestamp || !input.signature?.startsWith("v1=")) {
    return false;
  }
  const timestamp = Number(input.timestamp);
  const now = input.nowSeconds ?? Math.floor(Date.now() / 1000);
  if (
    !Number.isFinite(timestamp) ||
    Math.abs(now - timestamp) > (input.toleranceSeconds ?? 300)
  ) {
    return false;
  }
  const expected = createHmac("sha256", input.secret)
    .update(`${input.timestamp}.${input.rawBody}`)
    .digest("hex");
  const actual = input.signature.slice(3);
  const expectedBuffer = Buffer.from(expected, "hex");
  const actualBuffer = Buffer.from(actual, "hex");
  return (
    expectedBuffer.length === actualBuffer.length &&
    timingSafeEqual(expectedBuffer, actualBuffer)
  );
}
